libosmscout/src/osmscout/routing/RouteNodeDataFile.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+  This source is part of the libosmscout library
+  Copyright (C) 2018  Tim Teulings
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
+*/
+
+#include <osmscout/routing/RouteNodeDataFile.h>
+
+namespace osmscout {
+
+  RouteNodeDataFile::RouteNodeDataFile(const std::string& datafile,
+                           size_t cacheSize)
+  : datafile(datafile),
+    cache(cacheSize),
+    tileCalculator(0)
+  {
+  }
+
+  bool RouteNodeDataFile::Open(const TypeConfigRef& typeConfig,
+                         const std::string& path,
+                         bool memoryMappedData)
+  {
+    this->typeConfig=typeConfig;
+
+    datafilename=AppendFileToDir(path,datafile);
+
+    try {
+      FileOffset indexFileOffset;
+      uint32_t   dataCount;
+      uint32_t   indexEntryCount;
+      uint32_t   tileMag;
+
+      scanner.Open(datafilename,
+                   FileScanner::LowMemRandom,
+                   memoryMappedData);
+
+      scanner.Read(indexFileOffset);
+      scanner.Read(dataCount);
+      scanner.Read(tileMag);
+
+      tileCalculator=TileCalculator(std::pow(2,tileMag));
+
+      scanner.SetPos(indexFileOffset);
+      scanner.Read(indexEntryCount);
+
+      for (size_t i=1; i<=indexEntryCount; i++) {
+        Pixel      cell;
+        IndexEntry entry;
+
+        scanner.Read(cell.x);
```

Comment:
isn't this duplication of coverage.idx ?

Reply:
Yes and no. Both indexes can offer the same information: "Is the given tile covered by the database". But IMHO they still differ: The coverage.idx should be much smaller and faster. Since it maps its information onto a single bit (in memory). The index that is part of the router data file, holds information where tile data is places and how much entries it has. This means: The index is bigger in memory and lookup is slower (not bitmap lookup (O(1)) but a map (O(logn)).

If I need coverage for map, I wold possibly use coverage.idx, and the route data file only if I actively do routing.

The coverage.idx is pure optional, so it is up to the actual client, which index it wants to use.